### PR TITLE
ENH, BLD: Fix math feature detection for wasm 

### PR DESCRIPTION
**Pull request by hoodmane**
The web assembly linker is strict about function types, it is
unwilling to link a function defined with signature say double
f(double, double) to an invocation with signature void f(void). This
causes trouble in numpy/core/setup.py in the functions
check_math_capabilities and check_mathlib. 

This patch fixes the problem by giving config.try_link the correct
function signatures for these functions. In particular I added a
separate header file with all the math declarations. It would be be
possible to just include 'math.h' but we also need to parse the
declarations to figure out how to call the functions. If f has
arguments type1, type2, type3, we want to call it like f((type1)0,
(type2)0, (type3)0). Anyways it is easier to parse the arguments out
of our feature_detection_math.h than to worry about the possibility
that someone will have a differently formatted system math.h. We do a
test where we include both math.h and feature_detection_math.h to
ensure consistency between the signatures. 

I also separated out the fcntl functions, backtrace and madvise, and
strtold_l. This is because they require separate headers. strtold_l
requires testing both with the locale.h header and with the xlocale.h
header (this seems to vary even among different linuxes). All of the
functions I moved out of OPTIONAL_STDFUNCS are absent on windows and
some are absent on OSX so separating them out of the
OPTIONAL_STDFUNCS should mildly improve the speed of feature
detection on mac and windows (since they have all the functions
remaining in OPTIONAL_STDFUNCS).

### CODE REVIEW

File: numpy/core/setup_common.py [hunk 1]
```diff
@@ -104,6 +121,24 @@ def check_api_version(apiversion, codegen_dir):
...
 
+OPTIONAL_STDFUNCS_SIGS = {}
+OPTIONAL_STDFUNCS_CALLS = {}
+
+for func in OPTIONAL_STDFUNCS:
+    set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, func, "double", ["double"])
+
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "strtoull", "unsigned long long", ["const char*", "char**", "int"])
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "strtoll", "long long", ["const char*", "char**", "int"])
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "strtold_l", "long double", ["const char*", "char**", "int /* locale_t */"])
```
Comment (hoodmane) [inline numpy/core/setup_common.py:132]: `"int /* locale_t */"`
^^ It's a peculiarity of the way the wasm ABI handles structs that this works. If I'm allowed to `#include <xlocale.h>` then I would just use that.

File: numpy/core/setup_common.py [hunk 2]
```diff
@@ -104,6 +121,24 @@ def check_api_version(apiversion, codegen_dir):
...
+
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "pow", "double", ["double", "double"])
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "atan2", "double", ["double", "double"])
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "nextafter", "double", ["double", "double"])
+
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "madvise", "int", ["void *", "size_t", "int"])
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "backtrace", "int", ["void **", "int"])
+set_sig(OPTIONAL_STDFUNCS_SIGS, OPTIONAL_STDFUNCS_CALLS, "fallocate", "int", ["int", "int", "int /* off_t */", "int /* off_t */"])
```
Comment (hoodmane) [inline numpy/core/setup_common.py:140]: Similarly, I wasn't clear if `#include "fcntl.h"` is allowed. Surely if though `fcntl.h` is missing then so is `fallocate`?

File: numpy/core/setup.py [hunk 3]
```diff
@@ -129,33 +129,35 @@ def win32_checks(deflist):
         deflist.append('FORCE_NO_LONG_DOUBLE_FORMATTING')
 
 def check_math_capabilities(config, ext, moredefs, mathlibs):
-    def check_func(func_name):
+    def check_func(func_name, decl=True, call_args=True):
         return config.check_func(func_name, libraries=mathlibs,
-                                 decl=True, call=True)
+                                 decl=decl, call=True, call_args=call_args, headers=["stddef.h"])
```
Comment (hoodmane) [inline numpy/core/setup.py:134]: I hope it is okay to `#include "stddef.h"` for `size_t`?

**Comment by hoodmane** [comment]
It seems like `check_mathlib` actually *depends* on the signature for `exp` being wrong to work correctly? Very strange.

If this is `testa.c`:
```C
int exp (void); // some incorrect signature for `exp`
int main (void) {
  exp();
  return 0;
}
``` 
and `testb.c`:
```C
double exp (double); // the correct signature for `exp`
int main (void) {
  exp();
  return 0;
}
```
then `cc -c testa.c -o testa.o` followed by `cc testa.o -o testa` fails, but if you link with `-lm` the linking works. On the other hand, when declared with the correct signature, it will link even without `-lm`. 

Out of 
```py
MANDATORY_FUNCS = ["sin", "cos", "tan", "sinh", "cosh", "tanh", "fabs",
        "floor", "ceil", "sqrt", "log10", "log", "exp", "asin",
        "acos", "atan", "fmod", 'modf', 'frexp', 'ldexp']
```
if the wrong signatures are given, every single function requires `-lm` to link. However, with correct signatures, only `log` and `fmod` do. Does anyone understand this?!

**Comment by hoodmane** [comment]
On Mac, dynamic linking is failing with "Symbol not found: _npy_pow"
On windows, TestSpecialFloats.test_arcsinh and TestSpecialFloats.test_arccosh fail, they expected `nan` but get `+/-inf`.

I think the windows second failure suggests that on windows: 

1. for some reason my modification makes it detect that `asinh` and `acosh` aren't available
2. the fallback implementations of `acosh` and `asinh` handle `nan` slightly differently than the system math library implementations.

**Comment by mattip** [comment]
> Does anyone understand this?!

Not really. Does it depend on compiler/version of compiler? I find the Compiler Explorer https://godbolt.org/ can be useful for things like this.

**Comment by rgommers** [comment]
That old code hurts a little to look at. Changes are more verbose, but I don't see a problem merging them to improve WASM - it won't really make build system migration more difficult.

A few other thoughts since I didn't look at this code in ages:
- The whole math library checking business will be replaced by 3 lines of code at some point: https://mesonbuild.com/howtox.html#add-math-library-lm-portably. Pretty sure the Compaq Portable Math Library is no longer a thing we have to care about:)
- Some of the functions in `OPTIONAL_HEADERS` can probably become mandatory, we can rely on C99 support (except for complex support in a few cases).
- A lot of stuff like the IEEE macros check hasn't been touched in 8 years, it's going to be fun deciding whether to replicate that or throw it all away.

**Comment by h-vetinari** [comment]
> On Mac, dynamic linking is failing with "Symbol not found: _npy_pow"
> Cygwin also seems to be failing due to missing `npy_pow` symbol

Incidentally, I was also looking for `npy_pow` today, and couldn't find the source (commenting here because this is the only issue/PR that shows up by searching for it). I don't have full confidence that the GH code search finds all instances, but also locally I didn't get much more:
```
> findstr /L /S /N npy_pow numpy/*.*
numpy/core/include/numpy/npy_math.h:201:NPY_INPLACE double npy_pow(double x, double y);
numpy/core/include/numpy/npy_math.h:305:NPY_INPLACE float npy_powf(float x, float y);
numpy/core/include/numpy/npy_math.h:348:NPY_INPLACE npy_longdouble npy_powl(npy_longdouble x, npy_longdouble y);
numpy/core/src/npymath/npy_math_complex.c.src:422:/* private function for use in npy_pow{f, ,l} */
numpy/core/src/npymath/npy_math_internal.h.src:569:        return -npy_pow@c@(-x, 1. / 3.);
numpy/core/src/npymath/npy_math_internal.h.src:572:        return npy_pow@c@(x, 1. / 3.);
numpy/core/src/umath/scalarmath.c.src:419:    *out = npy_pow@c@(a, b);
numpy/core/src/umath/scalarmath.c.src:428:    const npy_float outf = npy_powf(af,bf);
```

**Comment by seberg** [comment]
@h-vetinari while there are some dummy [fallback definitions](https://github.com/numpy/numpy/blob/d94fc600c385de2bf77c65202d3cc80ce0f37a65/numpy/core/src/npymath/npy_math_internal.h.src#L427) (using double power), those are just aliases for the C power functions (they are C99, so they at least should always exist).  I bet we don't need them.  However, the ufunc, operator, and scalars may have their own paths (especially for integer powers) and to define fast-paths for certain powers.

**Comment by h-vetinari** [comment]
> However, the ufunc, operator, and scalars may have their own paths (especially for integer powers) and to define fast-paths for certain powers.

That's actually what I was trying to find out. I wanted to know how `power` is implemented to try to determine whether it would be more beneficial performance-wise to translate `power(2, x)` to `exp(log(2) * x)` (even though it's likely implemented like that...).

**Comment by seberg** [comment]
@h-vetinari there are some special paths for the operator `**`, trying the following: https://github.com/numpy/numpy/blob/7653829fa0d409d5850566cda1c031d9fadc1c22/numpy/core/src/multiarray/number.c#L449

The ufunc is "generated": https://github.com/numpy/numpy/blob/75edab9f7a7d95ecc62efc1b3b92642b6d45762d/numpy/core/code_generators/generate_umath.py#L399
That  may well be a bad idea: It will prevent any inlining of the `pow` function (or further Compiler optimizations), and just call the math libraries `pow`.  It may well be that this did not matter at the time it was written, but now would be much faster to write explicitly.  (The code does not generate a loop that would allow inlining, maybe that would be a solution as well: Change how the code generator works for `f=...`).

That is... If it is worthwhile.

**Comment by hoodmane** [comment]
Would appreciate if someone could review this @rgommers. We have merged exactly this patch into Pyodide 
https://github.com/pyodide/pyodide/pull/2256 
so in addition to passing the CI here, it also works correctly in wasm as I asserted (previously we had a hack to disable the feature detection and hard code the results).

**Review by mattip — APPROVED** [review]
LGTM. It always bothered me that we cannot build with `-Wall -Werror`.

**Comment by charris** [comment]
Could you add a note in `doc/release/upcoming_changes`? Look at other notes to see how that is done.

**Comment by hoodmane** [comment]
@charris How's this look?

File: doc/release/upcoming_changes/21154.improvement.rst [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+This is now fixed: we feature detect instead with code like:
+
+.. code-block:: C
+
+    double exp(double); // correct signature!
+    exp((double)0);
+
+which ``wasm-ld`` will link. As a side effect, the feature detection code now
+works without generating linker warnings.
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply doc/release/upcoming_changes/21154.improvement.rst]: Yeah brevity in technical writing is not my strongest point, I tend to get caught up in describing the details =)

[the remark this answers is not in the document]
Comment (hoodmane) [reply doc/release/upcoming_changes/21154.improvement.rst]: I substituted in your description, thanks.

**Comment by mattip** [comment]
[here](https://output.circle-artifacts.com/output/job/27143a33-f625-41b3-aa3d-3f78126639e8/artifacts/0/doc/build/html/release/1.23.0-notes.html#math-c-library-feature-detection-now-uses-correct-signatures) is the rendered note.

**Comment by mattip** [comment]
Flaky test_vector_matrix_values on win32 failure not connected to this PR. Merging

**Comment by seberg** [comment]
xref https://github.com/numpy/numpy/issues/21584 where these changes seem to cause some failure during detection of the required math functions (cos, etc.) on https://termux.com/.

**Comment by hoodmane** [comment]
I just read [this article](http://ewontfix.com/13/) which suggests feature detecting as follows:
```C
int main()
{
    double (*p)(const char *, char **, locale_t) = strtod_l;
}
```


Though this won't work if some toolchains define the function as a function-like macro.

**Comment by rgommers** [comment]
> Though this won't work if some toolchains define the function as a function-like macro.

The article seems to say that that's a feature - but it seems to me it's a problem, right?

**Comment by hoodmane** [comment]
To be fair I think neither the approach prior to this pr nor the one in this pr detect function-like macros. You would need to include the real headers for that. 

But I don't think the linked article is very clearly written -- it doesn't have any examples so it's hard to know in what edge cases these approaches behave differently.

**Pull request merged**